src/qobj/states.jl

```diff
@@ -32,15 +46,68 @@ It is also possible to specify the list of dimensions `dims` if different subsys
 basis(N::Int, pos::Int = 0; dims::Vector{Int} = [N]) = fock(N, pos, dims = dims)
 
 @doc raw"""
-    coherent(N::Real, α::Number)
+    coherent(N::Int, α::Number)
 
 Generates a coherent state ``\ket{\alpha}``, which is defined as an eigenvector of the bosonic annihilation operator ``\hat{a} \ket{\alpha} = \alpha \ket{\alpha}``.
 """
-function coherent(N::Real, α::T) where {T<:Number}
+function coherent(N::Int, α::T) where {T<:Number}
     a = destroy(N)
     return exp(α * a' - α' * a) * fock(N, 0)
 end
 
+@doc raw"""
+    fock_dm(N::Int, pos::Int=0; dims::Vector{Int}=[N], sparse::Bool=false)
+
+Density matrix representation of a Fock state.
+
+Constructed via outer product of [`fock`](@ref).
+"""
+function fock_dm(N::Int, pos::Int = 0; dims::Vector{Int} = [N], sparse::Bool = false)
+    ψ = fock(N, pos; dims = dims, sparse = sparse)
+    return ψ * ψ'
+end
+
+@doc raw"""
+    coherent_dm(N::Int, α::Number)
+
+Density matrix representation of a coherent state.
+
+Constructed via outer product of [`coherent`](@ref).
+"""
+function coherent_dm(N::Int, α::T) where {T<:Number}
+    ψ = coherent(N, α)
+    return ψ * ψ'
+end
+
+@doc raw"""
+    thermal_dm(N::Int, n::Real)
+
+Density matrix for a thermal state (generating thermal state probabilities) with the following arguments:
+- `N::Int`: Number of basis states in the Hilbert space
+- `n::Real`: Expectation value for number of particles in the thermal state.
+"""
+function thermal_dm(N::Int, n::Real)
+    β = log(1.0 / n + 1.0)
+    N_list = Array{Float64}(0:N-1)
+    data = exp.(-β .* N_list)
+    return QuantumObject(spdiagm(0 => data ./ sum(data)), Operator, [N])
```

Comment:
We should also have the case of both dense and sparse case.

Reply:
okay